tensorflow/core/transforms/remapper/pass.cc

```diff
@@ -306,6 +312,250 @@ class BasePatternActivationRewriter : public BasePatternRewriter {
   }
 };
 
+// NOTE(ezhulenev): See `BatchnormSpatialPersistentEnabled` documentation in the
+// `tensorflow/stream_executor/cuda/cuda_dnn.cc` for details.
+bool BatchnormSpatialPersistentEnabled() {
+#if CUDNN_VERSION >= 7402
+  static bool is_enabled = [] {
+    bool is_enabled = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar(
+        "TF_USE_CUDNN_BATCHNORM_SPATIAL_PERSISTENT",
+        /*default_val=*/false, &is_enabled));
+    return is_enabled;
+  }();
+  return is_enabled;
+#else
+  return false;
+#endif
+}
+
+// FusedBatchNorm[$is_training] + ... -> _FusedBatchNormEx[$is_training]
+//   (1) FusedBatchNorm + <Activation>
+//   (2) FusedBatchNorm + SideInput + <Activation>
+// only supported activation is Relu
+
+class FusedBatchNormExRewriter : public RemapperPatternBase {
+ public:
+  explicit FusedBatchNormExRewriter(OpPropertyHelper &helper)
+      : RemapperPatternBase("tfg.Relu", helper, PatternBenefit(1)) {}
+
+  // Constructor used by derived pattern rewritter class that may have
+  // different root operation name. Currently, pattern is
+  // matched from root op to its inputs.
+  explicit FusedBatchNormExRewriter(StringRef op_name, OpPropertyHelper &helper,
+                                    PatternBenefit benefit)
+      : RemapperPatternBase(op_name, helper, benefit) {}
+
+  using Pattern = FusedBatchNormEx;
+
+  bool is_valid_batch_norm(Operation *fused_batch_norm_op) const {
+    TFOp fusedbatchnormop_wrapper(fused_batch_norm_op);
+    if (!this->helper_.getDialect()->IsFusedBatchNorm(
+            fusedbatchnormop_wrapper)) {
+      return false;
+    }
+    // We fuse FusedBatchNorm on GPU or oneDNN CPU.
+    if (!this->helper_.isOneDNNEnabled() &&
+        !util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU)) {
+      return false;
+    }
+
+    TypeAttr attr = fused_batch_norm_op->getAttrOfType<TypeAttr>("T");
+    if (!attr) return false;
+    Type dtype_T = attr.getValue();
+
+    if (util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU)) {
+      // GPU supports float and half.
+      // Put this condition before check `isOneDNNEnabled()` because this node
+      // should be processed when it's on GPU and oneDNN CPU is enabled.
+      if (!dtype_T.isa<Float32Type, Float16Type>()) return false;
+    } else {
+      // Bfloat16 is available only with oneDNN.
+      // Half is not available with oneDNN.
+      if (this->helper_.isOneDNNEnabled() && !dtype_T.isa<Float32Type>() &&
+          !dtype_T.isa<BFloat16Type>()) {
+        return false;
+      }
+    }
+
+    // Get the FusedBatchNorm training mode.
+    auto training_attr =
+        fused_batch_norm_op->getAttrOfType<BoolAttr>("is_training");
+    if (!training_attr) return false;
+    auto is_training = training_attr.getValue();
```

Comment:
lit: It will be better to explicit write the type rather than using `auto` here

Reply:
Done